e2e-chatbot-app/app.py

```diff
@@ -0,0 +1,189 @@
+assert SERVING_ENDPOINT is not None, "SERVING_ENDPOINT must be set in app.yaml."
+
+ENDPOINT_SUPPORTS_FEEDBACK = endpoint_supports_feedback(SERVING_ENDPOINT)
+
+def reduce_chunks(chunks):
+    """
+    Reduce a list of ChatAgentChunk objects corresponding to a particular
+    message into a single ChatAgentMessage
+    """
+    deltas = [chunk.delta for chunk in chunks]
+    first_delta = deltas[0]
+    result_msg = first_delta
+    msg_contents = [first_delta.content]
+    for delta in deltas[1:]:
+        if delta.tool_calls:
+            result_msg = result_msg.model_copy(update={"tool_calls": delta.tool_calls})
+        if delta.tool_call_id:
+            result_msg = result_msg.model_copy(update={"tool_call_id": delta.tool_call_id})
+        msg_contents.append(delta.content)
+    result_msg = result_msg.model_copy(update={"content": "".join(msg_contents)})
+    return result_msg
+
+
+class Message(ABC):
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def to_input_messages(self):
+        """Convert this message into a list of dicts suitable for the model API."""
+        pass
+
+    @abstractmethod
+    def render(self, idx):
+        """Render the message in the Streamlit app."""
+        pass
+
+
+class UserMessage(Message):
+    def __init__(self, content):
+        super().__init__()
+        self.content = content
+
+    def to_input_messages(self):
+        return [{
+            "role": "user",
+            "content": self.content
+        }]
+
+    def render(self, idx):
+        with st.chat_message("user"):
+            st.markdown(self.content)
+
+
+class AssistantResponse(Message):
+    def __init__(self, messages, request_id):
+        super().__init__()
+        self.messages = messages
+        self.request_id = request_id
+
+    def to_input_messages(self):
+        return self.messages
+
+    def render(self, idx):
+        with st.chat_message("assistant"):
+            for msg in self.messages:
+                render_message(msg)
+
+            if self.request_id is not None:
+                render_assistant_message_feedback(idx, self.request_id)
+
+
+def render_message(msg):
+    if msg["role"] == "assistant" and "tool_calls" in msg:
+        for call in msg["tool_calls"]:
+            fn_name = call["function"]["name"]
+            args = call["function"]["arguments"]
+            st.markdown(msg["content"])
+            st.markdown(f"🛠️ Calling **`{fn_name}`** with:\n```json\n{args}\n```")
+    elif msg["role"] == "tool":
+        st.markdown("🧰 Tool Response:")
+        st.code(msg["content"], language="json")
+    elif msg["role"] == "assistant" and msg.get("content"):
+        st.markdown(msg["content"])
+
+# --- Init state ---
+if "history" not in st.session_state:
+    st.session_state.history = []
+
+st.title("🧱 Chatbot App")
+st.write(f"A basic chatbot using your own serving endpoint.")
+st.write(f"Endpoint name: `{SERVING_ENDPOINT}`")
+
+@st.fragment
```

Comment:
`st.fragment` is needed to ensure that updates to feedback (clicking +/-) don't trigger an app rerender (if the app re-renders midway through streaming output, the current partial streaming output/stream is lost)